language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionUtil.java

```diff
@@ -469,17 +457,236 @@ public static PositionDetails findCursorDetails(Range range, SyntaxTree syntaxTr
      * @param range {@link Range}
      * @return largest possible node
      */
-    private static NonTerminalNode largestExpressionNode(NonTerminalNode node, Range range) {
-        Predicate<NonTerminalNode> isWithinScope =
+    public static Node largestExpressionNode(Node node, Range range) {
+        Predicate<Node> isWithinScope =
                 tNode -> tNode != null && !(tNode instanceof ExpressionStatementNode) &&
                         CommonUtil.isWithinRange(CommonUtil.toPosition(tNode.lineRange().startLine()), range) &&
                         CommonUtil.isWithinRange(CommonUtil.toPosition(tNode.lineRange().endLine()), range);
         while (isWithinScope.test(node.parent())) {
             node = node.parent();
         }
+        if (node.kind() == SyntaxKind.ASSIGNMENT_STATEMENT) {
+            return ((AssignmentStatementNode) node).expression();
+        } else if (node.kind() == SyntaxKind.MODULE_VAR_DECL) {
+            return ((ModuleVariableDeclarationNode) node).typedBindingPattern().typeDescriptor();
+        } else if (node.kind() == SyntaxKind.LOCAL_VAR_DECL) {
+            return ((VariableDeclarationNode) node).typedBindingPattern().typeDescriptor();
+        }
         return node;
     }
 
+    /**
+     * Get the top level node type at the cursor line.
+     *
+     * @param position   {@link Position}
+     * @param syntaxTree {@link SyntaxTree}
+     * @return {@link String}   Top level node
+     */
+    public static Optional<Node> getTopLevelNode(Position position, SyntaxTree syntaxTree) {
+        NonTerminalNode member = CommonUtil.findNode(new Range(position, position), syntaxTree);
+        LinePosition cursorPosition = LinePosition.from(position.getLine(), position.getCharacter());
+        int cursorPosOffset = syntaxTree.textDocument().textPositionFrom(cursorPosition);
+        while (member != null) {
+            boolean isWithinStartSegment = isWithinStartCodeSegment(member, cursorPosOffset);
+            boolean isWithinBody = isWithinBody(member, cursorPosOffset);
+            if (!isWithinStartSegment && !isWithinBody) {
+                member = member.parent();
+                continue;
+            }
+
+            if (member.kind() == SyntaxKind.SERVICE_DECLARATION) {
```

Comment:
Following checks are bit hard to understand. Shall we try to simplify this? There are common conditions with the same return, we can merge them as well

Reply:
First part checks whether it is within `start segment` of the node. `isWithinStartSegment` has separate calculations base on the node-type. We can merge only variables, imports and assignments. Others are bit different.